std/parser.glu

```diff
@@ -1,295 +1,109 @@
-            | Some record ->
-                if predicate record.char then take_ record.rest
-                else if stream.start == stream2.start then
-                    Err { position = stream.start, message = "Unexpected token" }
-                else Ok {
-                            value = string.slice stream.buffer stream.start stream2.start,
-                            buffer = stream2,
-                        }
-            | None -> Ok {
-                        value = string.slice stream.buffer stream.start stream.end,
-                        buffer = stream2,
-                    }
-        take_ stream)
-
-/// Parses zero or more tokens passing `predicate` and returns the `String` between the start and
-/// end of those tokens
-let take predicate : (Char -> Bool) -> Parser String =
-    take1 predicate <|> wrap ""
-
-/// Parses using `p` and returns the `String` between the start and of what `p` parsed
-let recognize p : Parser a -> Parser String =
-    parser (\stream ->
-        match parser p stream with
-        | Ok a ->
-            Ok {
-                    value = string.slice stream.buffer stream.start a.buffer.start,
-                    buffer = a.buffer,
-                }
-        | Err err -> Err err)
-
-/// Skips over whitespace characters
-let spaces = take char.is_whitespace
-
-/// Creates a parser from a factory function. Useful to prevent mutually recursive parser from looping forever
-let lazy_parser f : (() -> Parser a) -> Parser a =
-    parser (\stream -> f () stream)
-
-/// Parses `x` between `l` and `r`, returning the result of `x`
-let between l r x : Parser a -> Parser b -> Parser c -> Parser c =
-    l *> x <* r
-
-rec
-/// Parses with `p` zero or more times
-let many p : Parser a -> Parser (List a) =
-    many1 p <|> wrap Nil
-
-/// Parses with `p` one or more times
-let many1 p : Parser a -> Parser (List a) =
-    do h = p
-    map (\t -> Cons h t) (many p)
-in
-rec
-/// Parses with `p` zero or more times, ignoring the result of the parser
-let skip_many p : Parser a -> Parser () =
-    skip_many1 p <|> wrap ()
-/// Parses with `p` one or more times, ignoring the result of the parser
-let skip_many1 p : Parser a -> Parser () =
-    seq p
-    skip_many p
-in
-/// Parses one of the characters of `s`
-let one_of s : String -> Parser Char =
-    satisfy (\first ->
-        let len = string.len s
-        let one_of_ i =
-            if i == len then
-                False
+let { Result } = import! std.result
+let streamlike @ { Streamlike, ? } = import! std.streamlike
+let string @ { String, ? } = import! std.string
+
+// TODO How handle atom streams generically? Stream-like interface?
+// TODO How handle error position generically? Zip stream with enumeration?
+type Parser s e a = StateT s (Result e) a
+// TODO How have Parser (Result Err) on surface but Streamer type (Option) internally?
+//   How handle custom error types this way? Do I need to build mini failure lib?
```

Comment:
`Streamlike` encodes the possibility of failure with `Option` and `Parser` does the same using `Result`. So, if `Parser` is built on `Streamlike`, there would need to be a way to convert between `Option` and `Result`.

Similarly, allowing the user to use a custom error type for `Parser` would require a way to translate the built-in errors to the custom error type.

In the comment, I'm considering whether or not it would be worthwhile to build a standard library module to handle such failure-case conversions, like `error-chain` and `failure` do in Rust.

For now, though, I'm thinking I might separate my definition of `Parser` from `Streamlike`, split `Streamlike`'s implementation into a different branch, and not implement user-defined errors in `Parser`. This would help to simplify `Parser`, like you suggested. I could then implement `Parser` so we could add these features back later without (or minimally) breaking the API.

Reply:
Aha, that makes sense 👍